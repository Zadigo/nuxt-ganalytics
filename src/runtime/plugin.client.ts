import { defineNuxtPlugin, useHead, useRouter, useRuntimeConfig } from '#app'
import { createGtm, type VueGtmUseOptions } from '@gtm-support/vue-gtm'

export default defineNuxtPlugin((nuxtApp) => {
  const router = useRouter()
  const config = useRuntimeConfig()
  const userOptions = nuxtApp.$config.public.ganalytics
  
  console.log('userOptions', userOptions)
  console.log('userOptions.gtm && userOptions.gtm.enabled', userOptions.gtm && userOptions.gtm.enabled)
  console.log('userOptions.gtm && userOptions.gtm.enabled', userOptions.gtm, userOptions.gtm?.enabled)

  if (userOptions.ga4 && userOptions.ga4.enabled) {
    const loadingStrategy = userOptions.ga4.loadingStrategy === 'async' ? 'async' : 'defer'
    const fullUrl = userOptions.ga4.url + `?id=${userOptions.ga4.id}`

    useHead({
      link: [
        {
          rel: 'preload',
          as: 'script',
          href: fullUrl
        }
      ],
      script: [
        {
          src: fullUrl,
          [loadingStrategy]: true,
          'data-ganalytics': ''
        }
      ]
    })
  }

  // console.log(defu(userOptions.gtm, ))
  if (userOptions.gtm && userOptions.gtm.enabled) {
    const options: VueGtmUseOptions = {
      ...userOptions.gtm,
      vueRouter: userOptions.gtm.enableRouterSync && router ? router as VueGtmUseOptions['vueRouter'] : undefined
    }

    console.log('userOptions.gtm', userOptions.gtm, options)
    nuxtApp.vueApp.use(createGtm(options))
  }
})
