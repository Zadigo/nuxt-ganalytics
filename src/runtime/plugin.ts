import { defineNuxtPlugin, useRouter } from '#app'
import { createGtm, type VueGtmUseOptions } from '@gtm-support/vue-gtm'

export default defineNuxtPlugin((nuxtApp) => {
  if (import.meta.client) {
    const router = useRouter()
    const userOptions = nuxtApp.$config.public.gtm
    const options: VueGtmUseOptions = {
      ...userOptions,
      vueRouter: userOptions.enableRouterSync && router ? router as VueGtmUseOptions['vueRouter'] : undefined
    }

    // Create a new GTM instance to be used
    nuxtApp.vueApp.use(createGtm(options))
  }
})
