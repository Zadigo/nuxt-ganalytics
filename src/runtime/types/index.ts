// import type { TagConfigurations } from './analytics'

export * from './analytics'

/**
 * @example
 * defineNuxtconfig({
 *  public: {
 *    ganalytics: {
 *        ga4: {
 *          id: ["consent", "default", {}]  
 *        } 
 *      }
 *    }
 * })
 */
// export interface TagModuleConfigurationOptions {
//   id: string
//   configuration: TagConfigurations['consent']
// }
