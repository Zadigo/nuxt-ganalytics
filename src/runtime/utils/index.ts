import type { DataLayerObject } from '@gtm-support/vue-gtm'
import type { RuntimeConfig } from 'nuxt/schema'
import type { Ref } from 'vue'
import { ref } from 'vue'
import type { CommandParameters, EventNames, GACommand, GAnalyticsDatalayerObject, GTMCommand } from '~/src/runtime/types'
import { defineCommand, defineConfig } from './payload'

export * from './payload'

/**
 * Pushes arguments to the existing datalayer container
 * @param payload The pyaload to be used in the layer
 */
export function dataLayerObject<T extends IArguments, R extends GAnalyticsDatalayerObject[] | DataLayerObject[]>(payload: T | undefined): R | undefined {
  if (import.meta.client && window.dataLayer && payload) {
    console.log('arguments', payload)
    window.dataLayer?.push(payload)
    // TODO: Maybe return a reactive object that 
    // contains the payload
    return Array.from(payload) as R
  }
}

/**
 * Checks if the datalayer contains a given tag name
 * @param name The name of the tag to check
 */
export function hasTag(name: GACommand | GTMCommand | string): boolean {
  if (import.meta.client && window.dataLayer) {
    // FIXME: Fix the type of window.dataLayer
    const results = window.dataLayer.filter(item => {
      const items = Array.from(item)
      return items.includes(name)
    })
    return results.length > 0
  } else {
    return false
  }
}

/**
 * Initializes GA4
 * @param config Nuxt runtime configuration
 * @example
 * gtag("js", new Date())
 * gtag("config", "G-123", {})
 */
export function initializeAnalytics(config: RuntimeConfig): Ref<boolean> {
  const stateCompleted = ref<boolean>(false)
  
  if (import.meta.client) {  
    window.dataLayer = window.dataLayer || []

    if (!hasTag('js')) {
      dataLayerObject(defineCommand('js', new Date()))
    }
  
    if (config.public.ganalytics.ga4) {
      const defaultParams: CommandParameters = {'debug': 'true'}
  
      if (!config.public.ganalytics.ga4.enableDebug) {
        delete defaultParams['debug']
      }
      
      const id = config.public.ganalytics.ga4.id
      if (typeof id === 'string') {
        if (!hasTag(id)) {
          dataLayerObject(defineConfig(id, defaultParams))
        }
      } else if (Array.isArray(id)) {
        id.forEach((item) => {
          if (typeof item === 'string') {
            if (!hasTag(item)) {
              dataLayerObject(defineConfig(item, defaultParams))
            }
          }
        })
      }
      stateCompleted.value = true
    }
  }
  
  // TODO: Maybe create a reactive object that returns the state and
  // the results from the datalayerObject function
  return stateCompleted
}
