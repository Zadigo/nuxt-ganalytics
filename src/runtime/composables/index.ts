import { useCookie } from '#app'
import { ref } from 'vue'
import { dataLayerObject, defineConsent } from '../utils'

import type { ConsentParameters, CustomGAnalyticsCookie } from '../types'

export * from './events'
export * from './tags'

export type ConsentArgs = keyof Omit<ConsentParameters, 'wait_for_update'>

/**
 * This composable provides methods to update consent parameters and manage user consent preferences
 */
export function useConsent() {
  if (!import.meta.client) {
    return {
      cookie: ref(null),
      updateConsent: () => {},
      acceptAll: () => {},
      denyAll: () => {}
    }
  }

  const defaultCookie = ref<CustomGAnalyticsCookie>({
    consent: {
      wait_for_update: 5000
    }
  })

  const cookie = useCookie<CustomGAnalyticsCookie | undefined>('ganalytics', { sameSite: 'strict', secure: true })

  async function updateConsent(params?: ConsentParameters) {
    if (params) {
      dataLayerObject(defineConsent(params, 'update'))

      if (!cookie.value) {
        cookie.value = defaultCookie.value
      }

      cookie.value.consent = params
    }
  }

  async function acceptAll(region?: string[]) {
    await updateConsent({
      ad_personalization: 'granted',
      ad_storage: 'granted',
      ad_user_data: 'granted',
      analytics_storage: 'granted',
      functionality_storage: 'granted',
      security_storage: 'granted',
      personalization_storage: 'granted',
      region
    })
  }
  
  async function denyAll(region?: string[]) {
    await updateConsent({
      ad_personalization: 'denied',
      ad_storage: 'denied',
      ad_user_data: 'denied',
      analytics_storage: 'denied',
      functionality_storage: 'denied',
      security_storage: 'denied',
      personalization_storage: 'denied',
      region
    })
  }

  return {
    /**
     * The cookie used to store consent parameters
     * for the user
     */
    cookie,
    /**
     * Acceps all consent for the user
     * @param region The region for which consent is accepted
     */
    acceptAll,
    /**
     * Accept or update consent parameters for the user
     * @param params The consent parameters to update
     */
    updateConsent,
    /**
     * Denies all consent for the user
     * @param region The region for which consent is denied
     */
    denyAll
  }
}
