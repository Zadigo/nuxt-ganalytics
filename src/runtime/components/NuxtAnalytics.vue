<template>
  <ClientOnly>
    <div ref="ganalyticsEl" data-id="ganalytics-container">
      <slot :attrs="defaultAttrs" />
    </div>
  </ClientOnly>
</template>

<script setup lang="ts">
import { useCounter, useDebounceFn } from '@vueuse/core'
import { computed, useTemplateRef } from 'vue'
import { useAnalyticsEvent } from '../composables'
import type { EventNames, EventParameters } from '../types'
import { dataLayerObject, defineEvent } from '../utils'

interface Props {
  event: EventNames
  params: EventParameters
  debounce?: number
}

const ganalyticsEl = useTemplateRef<HTMLElement>('ganalyticsEl')
const props = withDefaults (defineProps<Props>(), { debounce: 0 })
const emit = defineEmits<{ 'ga-event': [ data: ReturnType<typeof dataLayerObject>  | undefined ] }>()

const { sendEvent } = useAnalyticsEvent()
const { inc: increment, count } = useCounter()

/**
 * Method to send a Google Analytics event with 
 * the specified parameters
 */
async function sendTemplateEvent () {
  const parsedResult = sendEvent(defineEvent(props.event, props.params))
  increment(1)
  emit('ga-event', parsedResult)
}

const debouncedSendTemplateEvent = useDebounceFn(sendTemplateEvent, props.debounce)

const defaultAttrs = computed(() => {
  return {
    count,
    sendTemplateEvent: debouncedSendTemplateEvent
  }
})

// onMounted(() => {
//   console.log(ganalyticsEl.value)
//   if (ganalyticsEl.value) {
//     console.log(ganalyticsEl.value)
//     // const children = ganalyticsEl.value.children
//     // console.log('NuxtAnalytics', children)

//     // if (children.length > 1) {
//     //   console.warn('NuxtAnalytics: Only one child element is allowed inside the NuxtAnalytics component.')
//     // } else if (children.length === 1) {
//     //   const child = children[0]
//     //   if (child instanceof HTMLElement) {
//     //     child.setAttribute('data-id', 'ganalytics-item')
//     //   }
//     // } else {
//     //   console.warn('NuxtAnalytics: No child element found inside the NuxtAnalytics component.')
//     // }
//   }
// })
</script>
