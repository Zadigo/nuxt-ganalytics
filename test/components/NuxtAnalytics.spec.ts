import { mountSuspended } from '@nuxt/test-utils/runtime'
import { describe, expect, it, vi } from 'vitest'
import type { EventNames, EventParameters } from '../../src/module'
import NuxtAnalytics from '../../src/runtime/components/NuxtAnalytics.vue'


describe('NuxtAnalytics', () => {
  vi.mock('@vueuse/core', () => ({
    useCounter: vi.fn(() => 0),
    useDebounceFn: vi.fn((fn) => fn)
  }))

  vi.mock('../../src/runtime/utils', () => ({
    dataLayerObject: vi.fn(),
    defineEvent: vi.fn(),
    initializeAnalytics: vi.fn()
  }))

  const props: { event: EventNames, params: EventParameters } = { event: 'add_to_cart', params: { coupon: '1' } }
  
  it.each([
    // Props
    ['with event and params', { props }],
    ['with debounce', { props: { ...props, debounce: 1000 } }],
  ])('Renders correctly with %s', async (name, { props }) => {
    const result = await mountSuspended(NuxtAnalytics, { props })
    expect(result.html).toMatchSnapshot()
  })
})
