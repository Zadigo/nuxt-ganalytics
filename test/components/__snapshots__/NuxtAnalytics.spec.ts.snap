// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`NuxtAnalytics > Renders correctly with event 1`] = `[Function]`;
